File: 2.add-two-numbers.py
#
# @lc app=leetcode id=2 lang=python3
#
# [2] Add Two Numbers
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        ln = ListNode(0)
        current = ln
        tpsum = 0
        while l1 or l2 or tpsum:
            if l1:
                tpsum += l1.val
                l1 = l1.next
            if l2:
                tpsum += l2.val
                l2 = l2.next
            current.next = ListNode(tpsum%10)
            current = current.next
            tpsum = tpsum//10
        return ln.next

        
# @lc code=end

File: C#/2.add-two-numbers.cs
/*
 * @lc app=leetcode id=2 lang=csharp
 *
 * [2] Add Two Numbers
 */

// @lc code=start
// using System.Collections.Generic;
//  public class ListNode {
//      public int val;
//      public ListNode next;
//      public ListNode(int val=0, ListNode next=null) {
//          this.val = val;
//          this.next = next;
//      }
//  }

public class Solution
    {
        public ListNode AddTwoNumbers(ListNode l1, ListNode l2){
            ListNode head = new ListNode();
            var pointer = head;
            int curval = 0;
            while(l1 != null || l2 != null){
                curval = (l1 == null ? 0 : l1.val) + (l2 == null ? 0 : l2.val) + curval;
                pointer.next = new ListNode(curval % 10);
                pointer = pointer.next;
                // overflow decimal, like 12, we keep 1 for the next loop
                curval = curval / 10;
                // if next l1/l2 is not null, go to the next node
                l1 = l1?.next;
                l2 = l2?.next;
            }
            // if there is overflow left, add a node
            if(curval != 0){
                pointer.next = new ListNode(curval);
            }
            return head.next;
        }
    }
// @lc code=end

